### ISSUE TITLE
Fluent LinkTo

### ISSUE THREAD

**Issue by bencassie** [roadrunnerlenny/etlbox#33]
Hi,

Whilst testing out a data flow with many stages the thought occurred to me that it would be useful to return 'this' from LinkTo(). That way it would be possible to fluently link together the DataFlows perhaps making the flows a bit more readable?

Ben

**Comment by roadrunnerlenny** [comment]
Hi Ben,

thanks for you input! Just to clarify, so you would like something like this: 

```C#
var source = new DBSource(...);
var trans1 = new RowTransformation(...);
var trans2 = new RowTransformation(...);
var dest = new DBDestination(...);

source.LinkTo(trans1).LinkTo(trans2).LinkTo(dest);
```

Actually, the LinkTo would not return 'this', but the linked object (e.g., source.LinkTo(trans1) would return trans1).

Is this what you had in mind?

**Comment by bencassie** [comment]
Hi,

Yes that was what I was thinking about. It just feels like it'd read clearer.

The question though is how could it be represented in the fluent fashion when predicates and multiple destination are a factor. Other than restating the flow for the negative predicates sending filtered out stuff to VoidDestinations etc.

**Comment by bencassie** [comment]
Hi there,

I see you've been working on the fluent change today. 

I too have been looking at that change and I see our work is very similar! My branch is currently here - 

https://github.com/bencassie/etlbox/tree/feature/Interface-changes-squashed

**Comment by bencassie** [comment]
Hi,

I've added an overload to the LinkTo method to shortcut negating predicates with VoidDestinations. 

It has made it a lot easier to fluently express the dataflow without having to retain references so that you can negate them later. Obviously its not a panacea if multiple predicated links are required but in simple filter situations it has merit.

```C#
flow.LinkTo(priceTransform, x => x.SecurityId.HasValue, true)
    .LinkTo(priceTypeTransform, x => x.PriceSourceId.HasValue, true)
    .LinkTo(securityTypeTransform, x => x.PriceTypeID.HasValue, true)
    .LinkTo(MTDTransform)                    
    .LinkTo(destination);
```

**Comment by roadrunnerlenny** [comment]
Hi Ben,

just had a look at your branch and your implementations. First of all, thank you for your great support! 

I like the idea of having a "default" `VoidDestination` for inverted predicates, that actually would make the general implementation of predicates a lot easier. On the other hand, this won't work if the logic would still not cover all the records. But the other way of manually linking the VoidDestination will still be there, so it will be an enhancment.  

I will have a deeper look at your implementation probably on Monday, but I think I would like to have it merged. Before that I probably will also introduce some kind of Base Class for Transformations, so that there won't so much code replication for the different LinkTo(...) implementations. After that, I would like to pick the code manually from your branch and add it to the codebase.

**Comment by bencassie** [comment]
Fantastic - I agree with you on the transformations base class also.

Great news :)

**Comment by roadrunnerlenny** [comment]
Hi Ben, 

I was able to do some refactoring (creating a base class to inherit from wasn't so helpful, so I created a composite class that is used when linking the objects). 
I also added a new LinkTo-method:
```IDataFlowLinkSource<TOutput> LinkTo(IDataFlowLinkTarget<TOutput> target, Predicate<TOutput> rowsToKeep, Predicate<TOutput> rowsIntoVoid)```

As you can see, I did not add a negation of the predicate inside the method. This would hide some logic which actually should be displayed when linking the objects together. Instead, I decided to add a second parameter to the method, which accepts another predicate - the first one is for the rows of the data flow to keep, the second one defines which rows will be send into the Void. 

What do you think?

**Comment by bencassie** [comment]
I think that's a great solution!

**Comment by roadrunnerlenny** [comment]
Good to hear! I will release a new version then, probably tomorrow.

**Comment by roadrunnerlenny** [comment]
Just a quick update: I just released version 1.6.2 (https://www.nuget.org/packages/ETLBox/1.6.2) which supports the fluent LinkTo() notation.
